local module = {}

local IROMIRAM0 = {0x00000000, 0x1FFFFFFF, "IROMIRAM"} -- Boot area

local DRAM0 = {0x20000000, 0x3FFFFFFF, "DRAM0"} -- 512MB
local DRAM1 = {0x40000000, 0x7FFFFFFF, "DRAM1"} -- 1024MB

local SROMC0 = {0x80000000, 0x87FFFFFF, "SROMC0"}
local SROMC1 = {0x88000000, 0x8FFFFFFF, "SROMC1"}
local SROMC2 = {0x90000000, 0x97FFFFFF, "SROMC2"}
local SROMC3 = {0x98000000, 0x9FFFFFFF, "SROMC3"}
local SROMC4 = {0xA0000000, 0xA7FFFFFF, "SROMC4"}
local SROMC5 = {0xA8000000, 0xAFFFFFFF, "SROMC5"} -- SROMC 크기는 모두 128MB.

local NAND = {0xB0000000, 0xBFFFFFFF, "NAND"} -- 256MB

local LPASRAM = {0xC0000000, 0xCFFFFFFF, "LPASRAM"} -- Low Power Audio SRAM 256MB | 사실상 버퍼.

local IROM = {0xD0000000, 0xD000FFFF, "IROM"} -- 64KB
local Reserved = {0xD0010000, 0xD001FFFF, ""} -- 64KB
local IRAM = {0xD0020000, 0xD0037FFF, "IRAM"} -- 96KB
local DMZ_ROM = {0xD8000000, 0xDFFFFFFF, "DMS_ROM"} -- 128MB

-- local SFRS = {0xE0000000, 0xFFFFFFFF} -- MMIO 512MB

local MMIO = {
	{0xE0000000, 0xE00FFFFF, "CHIPID"},
	{0xE0100000, 0xE01FFFFF, "SYSCON"},
	{0xE0200000, 0xE02FFFFF, "GPIO"},
	{0xE0300000, 0xE03FFFFF, "AXI_DMA"},
	{0xE0400000, 0xE04FFFFF, "AXI_PSYS"},
	{0xE0500000, 0xE05FFFFF, "AXI_PSFR"},
	{0xE0600000, 0xE06FFFFF, "TZPC2"},
	{0xE0700000, 0xE07FFFFF, "IEM_APC"},
	{0xE0800000, 0xE08FFFFF, "IEM_IEC"},
	{0xE0900000, 0xE09FFFFF, "PDMA0"},
	{0xE0A00000, 0xE0AFFFFF, "PDMA1"},
	{0xE0D00000, 0xE0DFFFFF, "CORESIGHT"},
	{0xE0E00000, 0xE0EFFFFF, "SECKEY"},
	{0xE0F00000, 0xE0FFFFFF, "ASYNC_AUDIO_PSYS"},
	{0xE1100000, 0xE11FFFFF, "SPDIF"},
	{0xE1200000, 0xE12FFFFF, "PCM1"},
	{0xE1300000, 0xE13FFFFF, "SPI0"},
	{0xE1400000, 0xE14FFFFF, "SPI1"},
	{0xE1600000, 0xE16FFFFF, "KEYIF"},
	{0xE1700000, 0xE17FFFFF, "TSADC"},
	{0xE1800000, 0xE18FFFFF, "I2C0 General"},
	{0xE1A00000, 0xE1AFFFFF, "I2C2 PMIC"},
	{0xE1B00000, 0xE1BFFFFF, "HDMI_CEC"},
	{0xE1C00000, 0xE1CFFFFF, "TZPC3"},
	{0xE1D00000, 0xE1DFFFFF, "AXI_GSYS"},
	{0xE1F00000, 0xE1FFFFFF, "ASYNC_PSFR_AUDIO"},
	{0xE2100000, 0xE21FFFFF, "I2S1"},
	{0xE2200000, 0xE22FFFFF, "AC97"},
	{0xE2300000, 0xE23FFFFF, "PCM0"},
	{0xE2500000, 0xE25FFFFF, "PWN"},
	{0xE2600000, 0xE26FFFFF, "ST"},
	{0xE2700000, 0xE27FFFFF, "WDT"},
	{0xE2800000, 0xE28FFFFF, "RTC_APBIF"},
	{0xE2900000, 0xE29FFFFF, "UART"},
	{0xE8000000, 0xE80FFFFF, "SROMC"},
	{0xE8200000, 0xE82FFFFF, "CFCON"},
	{0xEA000000, 0xEA0FFFFF, "SECSS"},
	{0xEB000000, 0xEB0FFFFF, "SDMMC0"},
	{0xEB100000, 0xEB1FFFFF, "SDMMC1"},
	{0xEB200000, 0xEB2FFFFF, "SDMMC2"},
	{0xEB300000, 0xEB3FFFFF, "SDMMC3"},
	{0xEB400000, 0xEB4FFFFF, "TSI"},
	{0xEC000000, 0xEC0FFFFF, "USBOTG"},
	{0xEC100000, 0xEC1FFFFF, "USBOTG_PHY_CON"},
	{0xEC200000, 0xEC2FFFFF, "USBHOST_EHCI"},
	{0xEC300000, 0xEC3FFFFF, "USBHOST_OHCI"},
	{0xED000000, 0xED0FFFFF, "MODEM"},
	{0xED100000, 0xED1FFFFF, "HOST"},
	{0xEE000000, 0xEE8FFFFF, "AUDIO_SS"},
	{0xEE900000, 0xEE9FFFFF, "AUDIO_SS / ASS_DMA"},
	{0xEEA00000, 0xEEAFFFFF, "AUDIO_SS / ASS_IBUF0"},
	{0xEEB00000, 0xEEBFFFFF, "AUDIO_SS / ASS_IBUF1"},
	{0xEEC00000, 0xEECFFFFF, "AUDIO_SS / ASS_OBUF0"},
	{0xEED00000, 0xEEDFFFFF, "AUDIO_SS / ASS_OBUF1"},
	{0xEEE00000, 0xEEEFFFFF, "AUDIO_SS / ASS_APB"},
	{0xEEF00000, 0xEEFFFFFF, "AUDIO_SS / ASS_ODO"},
	{0xF0000000, 0xF00FFFFF, "DMC0_SFR"},
	{0xF1000000, 0xF10FFFFF, "AXI_MSYS"},
	{0xF1100000, 0xF11FFFFF, "AXI_MSFR"},
	{0xF1200000, 0xF12FFFFF, "AXI_VSYS"},
	{0xF1400000, 0xF14FFFFF, "DMC1_SFR"},
	{0xF1500000, 0xF15FFFFF, "TZPC0"},
	{0xF1600000, 0xF16FFFFF, "SDM"},
	{0xF1700000, 0xF17FFFFF, "MFC"},
	{0xF1800000, 0xF18FFFFF, "ASYNC_MFC_VSYS0"},
	{0xF1900000, 0xF19FFFFF, "ASYNC_MFC_VSYS1"},
	{0xF1A00000, 0xF1AFFFFF, "ASYNC_DSYS_MSYS0"},
	{0xF1B00000, 0xF1BFFFFF, "ASYNC_DSYS_MSYS1"},
	{0xF1C00000, 0xF1CFFFFF, "ASYNC_MSFR_DSFR"},
	{0xF1D00000, 0xF1DFFFFF, "ASYNC_MSFR_PSFR"},
	{0xF1E00000, 0xF1EFFFFF, "ASYNC_MSYS_DMC0"},
	{0xF1F00000, 0xF1FFFFFF, "ASYNC_MSFR_MPERI"},
	{0xF2000000, 0xF20FFFFF, "VIC0"},
	{0xF2100000, 0xF21FFFFF, "VIC1"},
	{0xF2200000, 0xF22FFFFF, "VIC2"},
	{0xF2300000, 0xF23FFFFF, "VIC3"},
	{0xF2800000, 0xF28FFFFF, "TZIC0"},
	{0xF2900000, 0xF29FFFFF, "TZIC1"},
	{0xF2A00000, 0xF2AFFFFF, "TZIC2"},
	{0xF2B00000, 0xF2BFFFFF, "TZIC3"},
	{0xF3000000, 0xF30FFFFF, "G3D"},
	{0xF8000000, 0xF80FFFFF, "FIMD"},
	{0xF9000000, 0xF90FFFFF, "TVENC"},
	{0xF9100000, 0xF91FFFFF, "VP"},
	{0xF9200000, 0xF92FFFFF, "MIXER"},
	{0xFA000000, 0xFA0FFFFF, "G2D"},
	{0xFA100000, 0xFA1FFFFF, "HDMI_LINK"},
	{0xFA200000, 0xFA2FFFFF, "SMDMA"},
	{0xFA300000, 0xFA3FFFFF, "ROT"},
	{0xFA400000, 0xFA4FFFFF, "AXI_LSYS"},
	{0xFA500000, 0xFA5FFFFF, "DSIM"},
	{0xFA600000, 0xFA6FFFFF, "CSIS"},
	{0xFA700000, 0xFA7FFFFF, "AXI_DSYS"},
	{0xFA800000, 0xFA8FFFFF, "AXI_DSFR"},
	{0xFA900000, 0xFA9FFFFF, "I2C_HDMI_PHY"},
	{0xFAA00000, 0xFAAFFFFF, "AXI_TSYS"},
	{0xFAB00000, 0xFABFFFFF, "I2C_HDMI_DDC"},
	{0xFAC00000, 0xFACFFFFF, "AXI_XSYS"},
	{0xFAD00000, 0xFADFFFFF, "TZPC1"},
	{0xFAF00000, 0xFAFFFFFF, "ASYSNC_PSYS_DSYS_u0"},
	{0xFB200000, 0xFB2FFFFF, "FIMC0"},
	{0xFB300000, 0xFB3FFFFF, "FIMC1"},
	{0xFB400000, 0xFB4FFFFF, "FIMC2"},
	{0xFB600000, 0xFB6FFFFF, "JPEG"}
}

local Memory_Map = {
	IROMIRAM0,
	DRAM0,
	DRAM1,
	SROMC0,
	SROMC1,
	SROMC2,
	SROMC3,
	SROMC4,
	SROMC5,
	NAND,
	LPASRAM,
	IROM,
	Reserved,
	IRAM,
	DMZ_ROM,
	-- SFRS,
	table.unpack(MMIO)
}

function module.AccessMMIO(ACCESS, ADDRESS, DATA) : number -- READ 0, WRITE 1
	for i=1, #Memory_Map do
		local Map = Memory_Map[i]
		if (ADDRESS >= Map[1] and ADDRESS <= Map[2]) then
			local DEVICE = require("./DEVICES/" .. Map[3] .. "/DEVICE")
			if (ACCESS == 1) then -- WRITE
				DEVICE.WRITE(ADDRESS - Map[1], DATA)
				return 0
			elseif (ACCESS == 0) then -- READ
				local DATA = DEVICE.READ(ADDRESS - Map[1])
				return DEVICE.READ(ADDRESS - Map[1])
			end
		end
	end
end

return module
